overlord/snapstate/snapstate.go

```diff
@@ -1980,17 +1989,156 @@ func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state
+	// (this may be a subset of refresh-candidates due to conflicts).
+	ar.Set("snaps", toUpdate)
 
 	return names, tss, nil
 }
 
+// autoRefreshPhase2 creates tasks for refreshing snaps from updates.
+func autoRefreshPhase2(st *state.State, updates []*refreshCandidate) ([]*state.TaskSet, error) {
+	refreshAll := true
+	fromChange := ""
+	flags := &Flags{IsAutoRefresh: true}
+	userID := 0
+
+	deviceCtx, err := DevicePastSeeding(st, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	toUpdate := make([]minimalInstallInfo, len(updates))
+	for i, up := range updates {
+		toUpdate[i] = up
+	}
+
+	tr := config.NewTransaction(st)
+	checkDiskSpaceRefresh, err := features.Flag(tr, features.CheckDiskSpaceRefresh)
+	if err != nil && !config.IsNoOption(err) {
+		return nil, err
+	}
+	if checkDiskSpaceRefresh {
+		// check if there is enough disk space for requested snap and its
+		// prerequisites.
+		totalSize, err := installSize(st, toUpdate, 0)
+		if err != nil {
+			return nil, err
+		}
+		requiredSpace := safetyMarginDiskSpace(totalSize)
+		path := dirs.SnapdStateDir(dirs.GlobalRootDir)
+		if err := osutilCheckFreeSpace(path, requiredSpace); err != nil {
+			snaps := make([]string, len(updates))
+			for i, up := range updates {
+				snaps[i] = up.InstanceName()
+			}
+			if _, ok := err.(*osutil.NotEnoughDiskSpaceError); ok {
+				return nil, &InsufficientSpaceError{
+					Path:       path,
+					Snaps:      snaps,
+					ChangeKind: "refresh",
+				}
+			}
+			return nil, err
+		}
+	}
+
+	newAutoAliases, mustPruneAutoAliases, transferTargets, err := autoAliasesUpdate(st, nil, toUpdate)
+	if err != nil {
+		return nil, err
+	}
+
+	tasksets := make([]*state.TaskSet, 0, len(updates)+2) // 1 for auto-aliases, 1 for re-refresh
+
+	var pruningAutoAliasesTs *state.TaskSet
+	if len(mustPruneAutoAliases) != 0 {
+		var err error
+		pruningAutoAliasesTs, err = applyAutoAliasesDelta(st, mustPruneAutoAliases, "prune", refreshAll, fromChange, func(snapName string, _ *state.TaskSet) {})
+		if err != nil {
+			return nil, err
+		}
+		tasksets = append(tasksets, pruningAutoAliasesTs)
+	}
+
+	// wait for the auto-alias prune tasks as needed
+	scheduleUpdate := func(snapName string, ts *state.TaskSet) {
+		if pruningAutoAliasesTs != nil && (mustPruneAutoAliases[snapName] != nil || transferTargets[snapName]) {
+			ts.WaitAll(pruningAutoAliasesTs)
+		}
+	}
+
+	// first snapd, core, bases, then rest
+	sort.Stable(byType(toUpdate))
+	prereqs := make(map[string]*state.TaskSet)
+	var kernelTs, gadgetTs *state.TaskSet
+
+	for _, up := range toUpdate {
+		var snapst SnapState
+		if err := Get(st, up.InstanceName(), &snapst); err != nil {
+			if err != state.ErrNoState {
+				return nil, err
+			}
+			logger.Noticef("cannot refresh snap %q, it is no longer installed", up.InstanceName())
+			continue
+		}
+
+		snapUserID, err := userIDForSnap(st, &snapst, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		snapsup := up.(*refreshCandidate).SnapSetup
+		snapsup.UserID = snapUserID
```

Comment:
relatedly, this is mostly but not quite refreshCandidate.SnapSetupForUpdate, though that could be adjusted